auth/authenticators/iam-assume-authenticator.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * (C) Copyright IBM Corp. 2024.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Authenticator } from './authenticator';
+import { IamAssumeTokenManager } from '../token-managers';
+import { IamRequestOptions, IamRequestBasedAuthenticator } from './iam-request-based-authenticator';
+
+/** Configuration options for IAM Assume authentication. */
+export interface Options extends IamRequestOptions {
+  /** The IAM api key */
+  apikey: string;
+
+  /**
+   * Specify exactly one of [iamProfileId, iamProfileCrn, or iamProfileName] to
+   * identify the trusted profile whose identity should be used. If iamProfileId
+   * or iamProfileCrn is used, the trusted profile must exist in the same account.
+   * If and only if IAMProfileName is used, then iamAccountId must also be
+   * specified to indicate the account that contains the trusted profile.
+   */
+  iamProfileId?: string;
+  iamProfileCrn?: string;
+  iamProfileName?: string;
+
+  /**
+   * If and only if iamProfileName is used to specify the trusted profile, then
+   * iamAccountId must also be specified to indicate the account that contains
+   * the trusted profile.
+   */
+  iamAccountId?: string;
+}
+
+/**
+ * The IamAssumeAuthenticator obtains an IAM access token using the IAM "get-token"
+ * operation's "assume" grant type. The authenticator obtains an initial IAM access
+ * token from a user-supplied apikey, then exchanges this initial IAM access token
+ * for another IAM access token that has "assumed the identity" of the specified
+ * trusted profile.
+ *
+ * The bearer token will be sent as an Authorization header in the form:
+ *
+ *      Authorization: Bearer \<bearer-token\>
+ */
+export class IamAssumeAuthenticator extends IamRequestBasedAuthenticator {
+  protected tokenManager: IamAssumeTokenManager;
```

Comment:
The way that the inheritance is set up in this core made it fairly easy to implement everything, because the shared IAM-related logic is all handled automatically.

I decided not to include any state in this class, because it didn't seem necessary and I wanted to keep it lightweight. We can revisit that decision if need be. It can also be added later in a compatible way, if need be.